add maxpool2D, add more tests, handle integer parameters for maxpool2D

migeed-z opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #61310
* #61265
* #61239
* #61238
* **#61188**

Differential Revision: [D29626303](https://our.internmc.facebook.com/intern/diff/D29626303)

migeed-z committed:
add maxpool2D, add more tests, handle integer parameters for maxpool2D

[ghstack-poisoned]

migeed-z committed:
Update on "add maxpool2D, add more tests, handle integer parameters for maxpool2D"




[ghstack-poisoned]

migeed-z committed:
Update on "add maxpool2D, add more tests, handle integer parameters for maxpool2D"




[ghstack-poisoned]

migeed-z committed:
Update on "add maxpool2D, add more tests, handle integer parameters for maxpool2D"




[ghstack-poisoned]

migeed-z committed:
Update on "add maxpool2D, add more tests, handle integer parameters for maxpool2D"




[ghstack-poisoned]

migeed-z committed:
Update on "add maxpool2D, add more tests, handle integer parameters for maxpool2D"




[ghstack-poisoned]

migeed-z committed:
Update on "add maxpool2D, add more tests, handle integer parameters for maxpool2D"




[ghstack-poisoned]

jamesr66a commented on torch/fx/experimental/graph_gradual_typechecker.py:
```diff
@@ -269,6 +283,50 @@ def relu_inference_rule(n: Node, op_type):
+
+
+@register_inference_rule(torch.nn.MaxPool2d)
+def maxpool2d_inference_rule(n: Node, op_type):
```
same here, rename `op_type`

jamesr66a commented on test/fx/test_gradual_type.py:
```diff
@@ -460,5 +460,57 @@ def forward(self, x: TensorType((2, 2, 4, 5))):
                 assert isinstance(n.type, TensorType)
                 assert torch.Size(n.type.__args__) == B.forward(torch.rand(2, 2, 4, 5)).size()
 
+    def test_type_check_conv2D_and_maxpool2d(self):
```
Same comment here as conv2d, if we can be more exhaustive about testing `stride`, `padding`, etc that would be good

jamesr66a reviewed (commented):
Looks pretty good!

migeed-z committed:
Update on "add maxpool2D, add more tests, handle integer parameters for maxpool2D"




[ghstack-poisoned]

migeed-z committed:
Update on "add maxpool2D, add more tests, handle integer parameters for maxpool2D"




[ghstack-poisoned]

migeed-z committed:
Update on "add maxpool2D, add more tests, handle integer parameters for maxpool2D"




[ghstack-poisoned]

migeed-z committed:
Update on "add maxpool2D, add more tests, handle integer parameters for maxpool2D"




[ghstack-poisoned]

migeed-z committed:
Update on "add maxpool2D, add more tests, handle integer parameters for maxpool2D"




[ghstack-poisoned]

migeed-z committed:
Update on "add maxpool2D, add more tests, handle integer parameters for maxpool2D"




[ghstack-poisoned]

migeed-z committed:
Update on "add maxpool2D, add more tests, handle integer parameters for maxpool2D"




[ghstack-poisoned]

migeed-z committed:
Update on "add maxpool2D, add more tests, handle integer parameters for maxpool2D"




[ghstack-poisoned]

migeed-z committed:
Update on "add maxpool2D, add more tests, handle integer parameters for maxpool2D"




[ghstack-poisoned]

migeed-z commented:
@migeed-z has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29626303).

jamesr66a commented on torch/fx/experimental/graph_gradual_typechecker.py:
```diff
@@ -277,6 +289,37 @@ def relu_inference_rule(n: Node, module_instance):
+    - Current node type is consistent with the output type we will calculate
+    - Input size matches output size and the last two dimensions of the output
+      are w_out and h_out. The remaining dimensions are the same as the input
+    - Our final result is the greatest lower bound of the output we calculate
```
Is this supposed to say "upper bound"?

jamesr66a reviewed (approved):
LGTM. I think you said you were going to add more tests, but I'm just going to preemptively stamp since they'll probably look the same as the conv2d tests

migeed-z committed:
Update on "add maxpool2D, add more tests, handle integer parameters for maxpool2D"


Differential Revision: [D29626303](https://our.internmc.facebook.com/intern/diff/D29626303)

[ghstack-poisoned]

migeed-z committed:
Update on "add maxpool2D, add more tests, handle integer parameters for maxpool2D"


Differential Revision: [D29626303](https://our.internmc.facebook.com/intern/diff/D29626303)

[ghstack-poisoned]

migeed-z committed:
Update on "add maxpool2D, add more tests, handle integer parameters for maxpool2D"


Differential Revision: [D29626303](https://our.internmc.facebook.com/intern/diff/D29626303)

[ghstack-poisoned]

migeed-z commented:
@migeed-z has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29626303).

migeed-z committed:
Update on "add maxpool2D, add more tests, handle integer parameters for maxpool2D"


Differential Revision: [D29626303](https://our.internmc.facebook.com/intern/diff/D29626303)

[ghstack-poisoned]

migeed-z commented:
@migeed-z has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29626303).

migeed-z commented:
@migeed-z has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29626303).

migeed-z committed:
Update on "add maxpool2D, add more tests, handle integer parameters for maxpool2D"


Differential Revision: [D29626303](https://our.internmc.facebook.com/intern/diff/D29626303)

[ghstack-poisoned]

migeed-z commented:
@migeed-z has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29626303).
